citest/azure_testing/az_contract.py

```diff
@@ -0,0 +1,135 @@
+      #self.filter_all_objects_to_observation(context, doc, observation)
+    except ValueError as vex:
+        error = 'Invalid JSON in response: %s' % str(az_response)
+        logging.getLogger(__name__).info('%s\n%s\n-----------------\n',
+                                         error, traceback.format_exc())
+        observation.add_error(JsonError(error, vex))
+        return []
+
+    return observation.objects
+
+class AzClauseBuilder(jc.ContractClauseBuilder):
+  """A ContractClause that facilitate observing the Azure state """
+
+  def __init__(self, title, az, retryable_for_secs=0, strict=False):
+    """Construct new clause.
+
+    Attributes:
+        title: The string title for the clause is only for reporting purposes.
+        az: The AzAgent to make the observation for the clause to verify.
+        retryable_for_secs: Number of seconds that observations can be retried
+                            if their verification initially fails.
+    strict: DEPRECATED flag indicating whether the clauses (added later)
+        must be true for all objects (strict) or at least one (not strict).
+        See ValueObservationVerifierBuilder for more information.
+        This is deprecated because in the future this should be on a per
+        constraint basis.
+    """
+    super(AzClauseBuilder, self).__init__(
+        title=title, retryable_for_secs=retryable_for_secs)
+    self.__az  = az
+    self.__strict = strict
+
+  def collect_resources(self, az_resource, command,
+                        args=None, filter=None,
+                        no_resources_ok=False):
+    """Collect the Azure resources of a particular type.
+
+    Attributes:
+        az_resource: The az resource module name we're looking in (e.g. 'vm')
+        command: The az command name to run (e.g. 'list')
+        args: An array of strings containing the remaining az command parameters.
+        filter: If provided, a filter to use for refining the collection.
+        no_resources_ok: Whether or not the resource is required.
+            If the resource is not required, 'resource not found' error is
+            considered successful.
+    """
+    args = args or []
+    cmd = self.__az.build_az_command_args(
+        az_resource, command, args)
+
+    self.observer = AzObjectObserver(self.__az, cmd)
+
+    if no_resources_ok:
+      error_verifier = cli_agent.CliAgentObservationFailureVerifier(
+          title='"Not Found" permitted.',
+          error_regex='(?:.* operation: Cannot find .*)|(?:.*\(.*NotFound\).*)')
+      disjunction_builder = jc.ObservationVerifierBuilder(
+          'Collect {0} or Not Found'.format(command))
+      disjunction_builder.append_verifier(error_verifier)
+
+      collect_builder = jc.ValueObservationVerifierBuilder(
+          'Collect {0}'.format(command), strict=self.__strict)
+      disjunction_builder.append_verifier_builder(
+          collect_builder, new_term=True)
+#      self.verifier_builder.append_verifier_builder(
```

Comment:
You want this. What this does is make this clause's verifier look like this:

disjunction_builder
|
+----- OR -------+
|  ......................... |
error_verifier ...  collect_builder

When you later return the collect_builder, the caller will add more constraints onto that. The effect is that when the clause validates itself, it will check for the error or all the constraints that the caller adds later.
When you leave this out, you arent assigning any verifier at all. The caller will be adding constraints to the result, but there is no association at all and the disjuction _builder (with the error verifier) is not referenced anywhere.